revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,180 @@
+)
+
+type Options struct {
+	CertChain       []*x509.Certificate
+	SigningTime     time.Time
+	HTTPClient      *http.Client
+	MergeErrorsFunc func(errors []error) error
+}
+
+const pkixNoCheckOID string = "1.3.6.1.5.5.7.48.1.5"
+const ocspMaxResponseSize int64 = 20480 //bytes
+
+func OCSPStatus(opts Options) error {
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	certResults := make([]error, len(opts.CertChain))
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go certOCSPStatus(cert, opts.CertChain[i+1], &certResults[i], &wg, opts)
+		} else {
+			// Last is root cert, which will never be revoked by OCSP
+			certResults[len(opts.CertChain)-1] = nil
+		}
+	}
+
+	wg.Wait()
+	return opts.MergeErrorsFunc(certResults)
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert *x509.Certificate, issuer *x509.Certificate, result *error, wg *sync.WaitGroup, opts Options) {
+	defer wg.Done()
+	var err error
+
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		(*result) = NoOCSPServerError{}
+		return
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		resp, err := ocspRequest(cert, issuer, server, opts)
+		if err != nil {
+			// If there is a server error, attempt all servers before determining what to return to the user
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		if opts.SigningTime.After(resp.NextUpdate) {
+			err = errors.New("expired OCSP response")
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		foundNoCheck := false
+		for _, extension := range resp.Extensions {
+			if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+				foundNoCheck = true
+			}
+		}
+		if !foundNoCheck {
+			// This will be ignored until CRL is implemented
+			// If it isn't found, CRL should be used to verify the OCSP response
+			// TODO: add CRL support
+			// https://github.com/notaryproject/notation-core-go/issues/125
+			serverErrs[serverIndex] = PKIXNoCheckError{}
+			continue
+		}
+
+		if resp.Status == ocsp.Revoked {
+			if time.Now().After(resp.RevokedAt) {
```

Comment:
the time here should be singing time aka signature generation time

Reply:
I think I changed the wrong one here. I'll update this to opts.SigningTime